### PR TITLE
[mlir][intrange] Use `nsw`,`nuw` flags in inference

### PR DESCRIPTION
**Pull request by ubfx**
This patch includes the "no signed wrap" and "no unsigned wrap" flags,
which can be used to annotate some Ops in the `arith` dialect and also
in LLVMIR, in the integer range inference.

The general approach is to use saturating arithmetic operations to infer
bounds which are assumed to not wrap and use overflowing arithmetic
operations in the normal case. If overflow is detected in the normal case,
special handling makes sure that we don't underestimate the result range.

Stacked on https://github.com/llvm/llvm-project/pull/92641

### CODE REVIEW

**Comment by Hardcode84** [comment]
Nice, I had this in plans (https://github.com/llvm/llvm-project/issues/77392) for a long time, but never had time to work on it. Thanks for tackling this.

File: mlir/lib/Dialect/Arith/IR/InferIntRangeInterfaceImpls.cpp [hunk 1]
```diff
@@ -19,6 +19,16 @@ using namespace mlir;
 using namespace mlir::arith;
 using namespace mlir::intrange;
 
+intrange::OverflowFlags
+convertArithOverflowFlags(arith::IntegerOverflowFlags flags) {
```
Comment (Hardcode84) [inline mlir/lib/Dialect/Arith/IR/InferIntRangeInterfaceImpls.cpp:23]: static

File: mlir/lib/Dialect/Index/IR/InferIntRangeInterfaceImpls.cpp [hunk 2]
```diff
@@ -44,19 +44,32 @@ void BoolConstantOp::inferResultRanges(ArrayRef<ConstantIntRanges> argRanges,
 // we take the 64-bit result).
 //===----------------------------------------------------------------------===//
 
+// Some arithmetic inference functions allow specifying special overflow / wrap
+// behavior. We do not require this for the IndexOps and use this helper to call
+// the inference function without any `OverflowFlags`.
+std::function<ConstantIntRanges(ArrayRef<ConstantIntRanges>)>
+inferWithoutOverflowFlags(InferRangeWithOvfFlagsFn inferWithOvfFn) {
```
Comment (Hardcode84) [inline mlir/lib/Dialect/Index/IR/InferIntRangeInterfaceImpls.cpp:51]: static

**Comment by ubfx** [comment]
> Nice, I had this in plans (#77392) for a long time, but never had time to work on it. Thanks for tackling this.

Thanks for the review, I'll post in the meta issue once this is merged

**Review by krzysz00 — APPROVED** [review]
Seems like a solid change, approved

**Review by Hardcode84 — APPROVED** [review]

**Pull request merged**